### PR TITLE
[DO NOT MERGE][Sparse] Support spsp_mul on CPU.

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description

This PR is one of the work items in #5368. 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by czkkkkkk** [comment]
@dgl-bot

File: python/dgl/sparse/elementwise_op_sp.py [hunk 1]
```diff
@@ -83,8 +90,8 @@ def sp_sub(A: SparseMatrix, B: SparseMatrix) -> SparseMatrix:
 def sp_mul(A: SparseMatrix, B: Union[SparseMatrix, Scalar]) -> SparseMatrix:
     """Elementwise multiplication
 
-    If :attr:`B` is a sparse matrix, both :attr:`A` and :attr:`B` must be
-    diagonal matrices.
+    Note that if both :attr:`A` and :attr:`B` are sparse matrices, it currently
+    only support multiplication on CPU.
```
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply python/dgl/sparse/elementwise_op_sp.py]: Not a regression. If both matrices are diagonal, it can run on GPU. I changed this description.

**Review by frozenbugs — APPROVED** [review]

File: dgl_sparse/src/cpu/matrix_ops_impl.cc [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+template <c10::DeviceType XPU, typename IdType>
+std::tuple<std::shared_ptr<COO>, torch::Tensor, torch::Tensor>
+COOIntersectionImpl(
+    const std::shared_ptr<COO> &lhs, const std::shared_ptr<COO> &rhs) {
+  const int64_t lhs_len = lhs->indices.size(1);
+  const int64_t rhs_len = rhs->indices.size(1);
+
+  torch::Tensor lhs_row = lhs->indices.index({0}).contiguous();
+  const IdType *lhs_row_ptr = static_cast<IdType *>(lhs_row.data_ptr());
```
Comment (jermainewang) [inline dgl_sparse/src/cpu/matrix_ops_impl.cc:21]: Could provide a utility function to get data pointers, e.g., `const IdType *lhs_row_ptr = utils::GetPtr<IdType>(lhs->indices.index({0}))`.

File: dgl_sparse/src/cpu/matrix_ops_impl.cc [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  std::vector<IdType> ret_lhs_indices, ret_rhs_indices;
+
+  auto pair_hash_fn = [](const std::pair<IdType, IdType> &pair) {
+    return std::hash<IdType>()(pair.first) ^ std::hash<IdType>()(pair.second);
+  };
+  std::unordered_map<std::pair<IdType, IdType>, IdType, decltype(pair_hash_fn)>
+      pair_map(lhs_len, pair_hash_fn);
+  for (int64_t k = 0; k < lhs_len; ++k) {
+    pair_map.emplace(std::make_pair(lhs_row_ptr[k], lhs_col_ptr[k]), k);
+  }
```
Comment (jermainewang) [inline dgl_sparse/src/cpu/matrix_ops_impl.cc:39]: If we don't consider operator efficiency. We can simply do the following (inspired by https://discuss.pytorch.org/t/intersection-between-to-vectors-tensors/50364/10?u=minjie_wang)

1. Hash the integer pairs to one integer (e.g., by `i * N + j` where `N` is the number of rows/columns).
2. Concatenate lhs and rhs into one vector.
3. Compute unique and the unique indices.
4. Unhash the integer to integer pairs and recover the indices.

This way you may get an operator that supports both CPU/GPU.

Comment (czkkkkkk) [reply dgl_sparse/src/cpu/matrix_ops_impl.cc:39]: It is doable and not very slow. I think we can use this implementation. Then do you think we need to keep the current code structure for adding a new operator?

**Comment by jermainewang** [comment]
A new implementation was merged in #5543 . We leave this PR as a future reference for adding custom C++ sparse operators.

**Comment by mufeili** [comment]
PR closed as it's only for reference.

**Pull request closed without merging**